### PR TITLE
Correctly handle limited sync responses by resetting the thread timeline

### PR DESCRIPTION
**Pull request by justjanne**
Type: Defect
Fixes: https://github.com/vector-im/element-web/issues/23952

---
Here's what your changelog entry will look like:

## 🐛 Bug Fixes
 * Correctly handle limited sync responses by resetting the thread timeline ([\#3056](https://github.com/matrix-org/matrix-js-sdk/pull/3056)). Fixes vector-im/element-web#23952. Contributed by @justjanne.

### CODE REVIEW

File: src/models/thread.ts [hunk 1]
```diff
@@ -358,6 +358,55 @@ export class Thread extends ReadReceipt<EmittedEvents, EventHandlerMap> {
...
+     * We try to prevent that by detecting such cases and resetting the live timeline accordingly.
+     * @private
+     */
+    private async tryResetLiveTimeline(): Promise<void> {
+        // if we're still syncing, let the sync complete first
+        if (!this.client.isInitialSyncComplete()) {
+            return;
+        }
+
+        const lastKnownEvent = this.events[this.events.length - 1];
```
Comment (gsouquet) [inline src/models/thread.ts:375]: This won't work, because the message in the bundled relationship of the root event is only going to be the last event with a `thread` relation type.
However the last known event in the timeline could be a reaction that is a second degree relation to a thread event

File: src/models/thread.ts [hunk 2]
```diff
@@ -358,6 +358,56 @@ export class Thread extends ReadReceipt<EmittedEvents, EventHandlerMap> {
         this.pendingReplyCount = pendingEvents.length;
     }
 
+    /**
+     * Resets the live timeline if we're desynced.
+     *
+     * During connection failures, sometimes not all sync state for a thread is saved.
+     * After a reconnect, the thread will be missing some messages.
+     * We try to prevent that by detecting such cases and resetting the live timeline accordingly.
```
Comment (gsouquet) [inline src/models/thread.ts:366]:
```suggestion
     * After connection failures, the sync loop will not always return all missed messages
     * the thread cou.d be missing some messages. We try to detect those and reset the 
     * live timeline accordingly.
     * Comparing the `lastEvent` advertised in the root event bundled relationship to the 
     * the known thread relation in the timeline is usually a good signal
```

File: src/models/thread.ts [hunk 3]
```diff
@@ -358,6 +358,56 @@ export class Thread extends ReadReceipt<EmittedEvents, EventHandlerMap> {
...
+        // If they have, abort and try again.
+        if (lastKnownEvent !== this.lastReply((ev) => ev.isRelation(THREAD_RELATION_TYPE.name))
+            || lastActualEvent !== this.lastEvent
+            || !this.client.isInitialSyncComplete()) {
+            return this.tryResetLiveTimeline();
+        }
+
+        logger.warn("Resetting live timeline for thread " + this.id + " due to a desync");
+
+        // This turns the old livetimeline into a regular timeline, moves it, and creates a new livetimeline for us
```
Comment (gsouquet) [inline src/models/thread.ts:403]:
```suggestion
        // This turns the old live timeline into a regular timeline, moves it, and creates a new live timeline for us
```

**Comment by justjanne** [comment]
For this to be ready, we need https://github.com/matrix-org/synapse/issues/14830 to be resolved first.

**Comment by t3chguy** [comment]
@justjanne does that mean we also need to wait for a Synapse release and sufficient roll-out or?

**Comment by justjanne** [comment]
@t3chguy That, as well as whether we can solve this in alternative ways, is what I'll try to figure out today. I suspect we'll ship a workaround for now with a TODO to remove it at a later time.

**Review by gsouquet — CHANGES_REQUESTED** [review]
Code looks good ! But the tests are quite hard to follow as the code is not really documented and the test case names are not very explanatory

**Review by gsouquet — APPROVED** [review]
Looks good to me 👍
